From c343e41cae9eea75048a12f56b70c0fd8ee3d7ab Mon Sep 17 00:00:00 2001
From: Bo Yang <bo.yang@amlogic.com>
Date: Tue, 14 Jul 2015 20:21:59 +0800
Subject: [PATCH 5702/5965] PD#108759: tvout: change to platform driver to
 support dtd

To support dtd, we changed the char device driver to
platform device driver.

Use the old device driver name defined by TV_CLASS_NAME.

for tvout, use following dtd configuration for example:
	tvout {
		compatible = "amlogic,tvout";
		...
	}

for tvout2, use the following configuration for example:
	tvout2 {
        	compatible = "amlogic,tvout2";
        	...
	}

Change-Id: I7ca5be08c24fe6f06ace2feeb8ec3969af101e9e
---
 drivers/amlogic/display/vout/tvconf.c   |  96 ++-
 drivers/amlogic/display/vout2/tvconf2.c | 951 ++++++++++++------------
 2 files changed, 565 insertions(+), 482 deletions(-)

diff --git a/drivers/amlogic/display/vout/tvconf.c b/drivers/amlogic/display/vout/tvconf.c
index e4b1ce9bf8fb..f888e7c19fd6 100644
--- a/drivers/amlogic/display/vout/tvconf.c
+++ b/drivers/amlogic/display/vout/tvconf.c
@@ -19,8 +19,8 @@
  * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA
  *
  * Author:   jianfeng_wang@amlogic
- *		   
- *		   
+ *
+ *
  */
 
 #include <linux/version.h>
@@ -1190,7 +1190,7 @@ static int framerate_automation_process(int duration)
 
 	fr_vsource = get_vsource_frame_rate(duration);
 	fr_target = get_target_frame_rate(fr_vsource, policy);
-	
+
 	pvinfo = tv_get_current_info();
 	if( (pvinfo->sync_duration_num==fr_target) || (pvinfo->sync_duration_num==(fr_target/100)) )
 		return 0;
@@ -1206,7 +1206,7 @@ static int framerate_automation_process(int duration)
 	}
 	printk("%s[%d] fps_playing_flag = %d\n", __FUNCTION__,__LINE__, fps_playing_flag);
 	mode_target = get_target_vmode(fr_target);
-	
+
 	framerate_automation_set_mode(mode_target);
 
 	return 0;
@@ -1272,10 +1272,10 @@ static int tv_resume(int pm_event)
 	tv_set_current_vmode(info->vinfo->mode);
 	return 0;
 }
-#endif 
+#endif
 static vout_server_t tv_server={
 	.name = "vout_tv_server",
-	.op = {	
+	.op = {
 		.get_vinfo=tv_get_current_info,
 		.set_vmode=tv_set_current_vmode,
 		.validate_vmode=tv_validate_vmode,
@@ -1283,10 +1283,10 @@ static vout_server_t tv_server={
 		.disable = tv_module_disable,
 		.set_vframe_rate_hint = tv_set_vframe_rate_hint,
 		.set_vframe_rate_end_hint = tv_set_vframe_rate_end_hint,
-#ifdef  CONFIG_PM  
+#ifdef  CONFIG_PM
 		.vout_suspend=tv_suspend,
 		.vout_resume=tv_resume,
-#endif	
+#endif
 	},
 };
 
@@ -1306,15 +1306,15 @@ static void _init_vout(void)
 **	The 6th digit control s-video chroma output DAC number
 ** 	examble :
 **		echo  120120 > /sys/class/display/vdac_setting
-**		the first digit from the left side .	
+**		the first digit from the left side .
 ******************************************************/
-static void  parse_vdac_setting(char *para) 
+static void  parse_vdac_setting(char *para)
 {
 	int  i;
 	char  *pt=strstrip(para);
 	int len=strlen(pt);
 	u32  vdac_sequence=get_current_vdac_setting();
-	
+
 	amlog_mask_level(LOG_MASK_PARA,LOG_LEVEL_LOW,"origin vdac setting:0x%x,strlen:%d\n",vdac_sequence,len);
 	if(len!=6)
 	{
@@ -1329,7 +1329,7 @@ static void  parse_vdac_setting(char *para)
 		pt++;
 	}
 	amlog_mask_level(LOG_MASK_PARA,LOG_LEVEL_LOW,"current vdac setting:0x%x\n",vdac_sequence);
-	
+
 	change_vdac_setting(vdac_sequence,get_current_vmode());
 }
 
@@ -1432,7 +1432,7 @@ static struct syscore_ops tvconf_ops = {
 };
 #endif
 
-static int __init tv_init_module(void)
+static int tvout_probe(struct platform_device *pdev)
 {
 	int  ret ;
 
@@ -1441,7 +1441,7 @@ static int __init tv_init_module(void)
 	register_syscore_ops(&tvconf_ops);
 #endif
 
-	info=&disp_module_info;
+	info = &disp_module_info;
 	printk("%s\n", __func__);
 
 	/*if (!info)
@@ -1449,12 +1449,12 @@ static int __init tv_init_module(void)
 		amlog_mask_level(LOG_MASK_INIT,LOG_LEVEL_HIGH,"can't alloc display info struct\n");
 		return -ENOMEM;
 	}*/
-	
+
 	memset(info, 0, sizeof(disp_module_info_t));
 
 	sprintf(info->name,TV_CLASS_NAME) ;
 	ret=register_chrdev(0,info->name,&am_tv_fops);
-	if(ret <0) 
+	if (ret < 0)
 	{
 		amlog_mask_level(LOG_MASK_INIT,LOG_LEVEL_HIGH,"register char dev tv error\n");
 		return  ret ;
@@ -1462,7 +1462,7 @@ static int __init tv_init_module(void)
 	info->major=ret;
 	_init_vout();
 	amlog_mask_level(LOG_MASK_INIT,LOG_LEVEL_HIGH,"major number %d for disp\n",ret);
-	if(vout_register_server(&tv_server))
+	if (vout_register_server(&tv_server))
 	{
 		amlog_mask_level(LOG_MASK_INIT,LOG_LEVEL_HIGH,"register tv module server fail\n");
 	}
@@ -1474,27 +1474,30 @@ static int __init tv_init_module(void)
 	return 0;
 
 }
-static __exit void tv_exit_module(void)
+
+
+static int __exit tvout_remove(struct platform_device *pdev)
 {
 	int i;
-	
-	if(info->base_class)
+
+	if (info->base_class)
 	{
-		for(i=0;i<ARRAY_SIZE(tv_attr);i++)
+		for (i=0; i<ARRAY_SIZE(tv_attr); i++)
 		{
 			class_remove_file(info->base_class,tv_attr[i]) ;
 		}
-		
+
 		class_destroy(info->base_class);
-	}	
-	if(info)
+	}
+	if (info)
 	{
 		unregister_chrdev(info->major,info->name)	;
 		//kfree(info);
 	}
 	vout_unregister_server(&tv_server);
-	
+
 	amlog_mask_level(LOG_MASK_INIT,LOG_LEVEL_HIGH,"exit tv module\n");
+	return 0;
 }
 
 #if (MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8)
@@ -1528,8 +1531,47 @@ static int __init cvbs_performance_setup(char* line)
 }
 __setup("cvbsdrv=", cvbs_performance_setup);
 
-arch_initcall(tv_init_module);
-module_exit(tv_exit_module);
+
+#ifdef CONFIG_USE_OF
+static const struct of_device_id tvout_dt_match[] = {
+	{
+		.compatible = "amlogic,tvout",
+	},
+	{},
+};
+#endif
+
+static struct platform_driver tvout_driver = {
+	.driver = {
+		.name = TV_CLASS_NAME,
+		.owner = THIS_MODULE,
+#ifdef CONFIG_USE_OF
+		.of_match_table = tvout_dt_match,
+#endif
+	},
+	.probe = tvout_probe,
+	.remove = __exit_p(tvout_remove),
+};
+
+static int __init tvout_init_module(void)
+{
+	printk("%s module init\n", __func__);
+	if (platform_driver_register(&tvout_driver)) {
+		printk("%s failed to register module\n", __func__);
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static void __exit tvout_exit_module(void)
+{
+	printk("%s module exit\n", __func__);
+	platform_driver_unregister(&tvout_driver);
+}
+
+arch_initcall(tvout_init_module);
+module_exit(tvout_exit_module);
 
 MODULE_DESCRIPTION("display configure  module");
 MODULE_LICENSE("GPL");
diff --git a/drivers/amlogic/display/vout2/tvconf2.c b/drivers/amlogic/display/vout2/tvconf2.c
index b63c54ef23db..51297e7b1570 100755
--- a/drivers/amlogic/display/vout2/tvconf2.c
+++ b/drivers/amlogic/display/vout2/tvconf2.c
@@ -19,8 +19,8 @@
  * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA
  *
  * Author:   jianfeng_wang@amlogic
- *		   
- *		   
+ *
+ *
  */
 
 #include <linux/version.h>
@@ -70,471 +70,471 @@ SET_TV2_CLASS_ATTR(vdac_setting,parse_vdac_setting)
 
 static const tvmode_t vmode_tvmode_tab[] =
 {
-    TVMODE_480I, TVMODE_480I_RPT, TVMODE_480CVBS, TVMODE_480P,
+	TVMODE_480I, TVMODE_480I_RPT, TVMODE_480CVBS, TVMODE_480P,
 #ifdef CONFIG_AML_VOUT_FRAMERATE_AUTOMATION
-    TVMODE_480P_59HZ,
+	TVMODE_480P_59HZ,
 #endif
-    TVMODE_480P_RPT, TVMODE_576I, TVMODE_576I_RPT, TVMODE_576CVBS, TVMODE_576P, TVMODE_576P_RPT, TVMODE_720P,
+	TVMODE_480P_RPT, TVMODE_576I, TVMODE_576I_RPT, TVMODE_576CVBS, TVMODE_576P, TVMODE_576P_RPT, TVMODE_720P,
 #ifdef CONFIG_AML_VOUT_FRAMERATE_AUTOMATION
-    TVMODE_720P_59HZ , // for 720p 59.94hz
+	TVMODE_720P_59HZ , // for 720p 59.94hz
 #endif
-    TVMODE_1080I,
+	TVMODE_1080I,
 #ifdef CONFIG_AML_VOUT_FRAMERATE_AUTOMATION
-    TVMODE_1080I_59HZ,
+	TVMODE_1080I_59HZ,
 #endif
-    TVMODE_1080P,
+	TVMODE_1080P,
 #ifdef CONFIG_AML_VOUT_FRAMERATE_AUTOMATION
-    TVMODE_1080P_59HZ , // for 1080p 59.94hz
+	TVMODE_1080P_59HZ , // for 1080p 59.94hz
 #endif
-    TVMODE_720P_50HZ, TVMODE_1080I_50HZ, TVMODE_1080P_50HZ,TVMODE_1080P_24HZ, 
+	TVMODE_720P_50HZ, TVMODE_1080I_50HZ, TVMODE_1080P_50HZ,TVMODE_1080P_24HZ,
 #ifdef CONFIG_AML_VOUT_FRAMERATE_AUTOMATION
-    TVMODE_1080P_23HZ , // for 1080p 23.97hz
+	TVMODE_1080P_23HZ , // for 1080p 23.97hz
 #endif
-    TVMODE_4K2K_30HZ,
+	TVMODE_4K2K_30HZ,
 #ifdef CONFIG_AML_VOUT_FRAMERATE_AUTOMATION
-    TVMODE_4K2K_29HZ , // for 4k2k 29.97hz
+	TVMODE_4K2K_29HZ , // for 4k2k 29.97hz
 #endif
-    TVMODE_4K2K_25HZ, TVMODE_4K2K_24HZ,
+	TVMODE_4K2K_25HZ, TVMODE_4K2K_24HZ,
 #ifdef CONFIG_AML_VOUT_FRAMERATE_AUTOMATION
-    TVMODE_4K2K_23HZ , // for 4k2k 23.97hz
+	TVMODE_4K2K_23HZ , // for 4k2k 23.97hz
 #endif
-    TVMODE_4K2K_SMPTE,
-    TVMODE_VGA, TVMODE_SVGA, TVMODE_XGA, TVMODE_SXGA
+	TVMODE_4K2K_SMPTE,
+	TVMODE_VGA, TVMODE_SVGA, TVMODE_XGA, TVMODE_SXGA
 };
 
 
-static const vinfo_t tv_info[] = 
+static const vinfo_t tv_info[] =
 {
-    { /* VMODE_480I */
-        .name              = "480i",
+	{ /* VMODE_480I */
+	.name              = "480i",
 	.mode              = VMODE_480I,
-        .width             = 720,
-        .height            = 480,
-        .field_height      = 240,
-        .aspect_ratio_num  = 4,
-        .aspect_ratio_den  = 3,
-        .sync_duration_num = 60,
-        .sync_duration_den = 1,
-        .video_clk         = 27000000,
-    },
+	.width             = 720,
+	.height            = 480,
+	.field_height      = 240,
+	.aspect_ratio_num  = 4,
+	.aspect_ratio_den  = 3,
+	.sync_duration_num = 60,
+	.sync_duration_den = 1,
+	.video_clk         = 27000000,
+	},
 #ifdef CONFIG_AML_VOUT_FRAMERATE_AUTOMATION
-    { /* VMODE_480P_59HZ */
-        .name              = "480p59hz",
-        .mode              = VMODE_480P_59HZ,
-        .width             = 720,
-        .height            = 480,
-        .field_height      = 480,
-        .aspect_ratio_num  = 4,
-        .aspect_ratio_den  = 3,
-        .sync_duration_num = 60000,
-        .sync_duration_den = 1001,
-        .video_clk         = 27000000,
-    },
+	{ /* VMODE_480P_59HZ */
+	.name              = "480p59hz",
+	.mode              = VMODE_480P_59HZ,
+	.width             = 720,
+	.height            = 480,
+	.field_height      = 480,
+	.aspect_ratio_num  = 4,
+	.aspect_ratio_den  = 3,
+	.sync_duration_num = 60000,
+	.sync_duration_den = 1001,
+	.video_clk         = 27000000,
+	},
 #endif
-    { /* VMODE_480I_RPT */
-        .name              = "480i_rpt",
-        .mode              = VMODE_480I_RPT,
-        .width             = 720,
-        .height            = 480,
-        .field_height      = 240,
-        .aspect_ratio_num  = 4,
-        .aspect_ratio_den  = 3,
-        .sync_duration_num = 60,
-        .sync_duration_den = 1,
-        .video_clk         = 27000000,
-     },
-     { /* VMODE_480CVBS*/
-        .name              = "480cvbs",
-        .mode              = VMODE_480CVBS,
-        .width             = 720,
-        .height            = 480,
-        .field_height      = 240,
-        .aspect_ratio_num  = 4,
-        .aspect_ratio_den  = 3,
-        .sync_duration_num = 60,
-        .sync_duration_den = 1,
-        .video_clk         = 27000000,
-    },
-    { /* VMODE_480P */
-        .name              = "480p",
-        .mode              = VMODE_480P,
-        .width             = 720,
-        .height            = 480,
-        .field_height      = 480,
-        .aspect_ratio_num  = 4,
-        .aspect_ratio_den  = 3,
-        .sync_duration_num = 60,
-        .sync_duration_den = 1,
-        .video_clk         = 27000000,
-    },
-    { /* VMODE_480P_RPT */
-        .name              = "480p_rpt",
-        .mode              = VMODE_480P_RPT,
-        .width             = 720,
-        .height            = 480,
-        .field_height      = 480,
-        .aspect_ratio_num  = 4,
-        .aspect_ratio_den  = 3,
-        .sync_duration_num = 60,
-        .sync_duration_den = 1,
-        .video_clk         = 27000000,
-    },
-    { /* VMODE_576I */
-        .name              = "576i",
-        .mode              = VMODE_576I,
-        .width             = 720,
-        .height            = 576,
-        .field_height      = 288,
-        .aspect_ratio_num  = 4,
-        .aspect_ratio_den  = 3,
-        .sync_duration_num = 50,
-        .sync_duration_den = 1,
-        .video_clk         = 27000000,
-    },
-    { /* VMODE_576I_RPT */
-        .name              = "576i_rpt",
-        .mode              = VMODE_576I_RPT,
-        .width             = 720,
-        .height            = 576,
-        .field_height      = 288,
-        .aspect_ratio_num  = 4,
-        .aspect_ratio_den  = 3,
-        .sync_duration_num = 50,
-        .sync_duration_den = 1,
-        .video_clk         = 27000000,
-    },
-    { /* VMODE_576I */
-        .name              = "576cvbs",
-        .mode              = VMODE_576CVBS,
-        .width             = 720,
-        .height            = 576,
-        .field_height      = 288,
-        .aspect_ratio_num  = 4,
-        .aspect_ratio_den  = 3,
-        .sync_duration_num = 50,
-        .sync_duration_den = 1,
-        .video_clk         = 27000000,
-    },
-    { /* VMODE_576P */
-        .name              = "576p",
-        .mode              = VMODE_576P,
-        .width             = 720,
-        .height            = 576,
-        .field_height      = 576,
-        .aspect_ratio_num  = 4,
-        .aspect_ratio_den  = 3,
-        .sync_duration_num = 50,
-        .sync_duration_den = 1,
-        .video_clk         = 27000000,
-    },
-    { /* VMODE_576P_RPT */
-        .name              = "576p_rpt",
-        .mode              = VMODE_576P_RPT,
-        .width             = 720,
-        .height            = 576,
-        .field_height      = 576,
-        .aspect_ratio_num  = 4,
-        .aspect_ratio_den  = 3,
-        .sync_duration_num = 50,
-        .sync_duration_den = 1,
-        .video_clk         = 27000000,
-    },
-    { /* VMODE_720P */
-        .name              = "720p",
-        .mode              = VMODE_720P,
-        .width             = 1280,
-        .height            = 720,
-        .field_height      = 720,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 60,
-        .sync_duration_den = 1,
-        .video_clk         = 74250000,
-    },
+	{ /* VMODE_480I_RPT */
+	.name              = "480i_rpt",
+	.mode              = VMODE_480I_RPT,
+	.width             = 720,
+	.height            = 480,
+	.field_height      = 240,
+	.aspect_ratio_num  = 4,
+	.aspect_ratio_den  = 3,
+	.sync_duration_num = 60,
+	.sync_duration_den = 1,
+	.video_clk         = 27000000,
+	},
+	{ /* VMODE_480CVBS*/
+	.name              = "480cvbs",
+	.mode              = VMODE_480CVBS,
+	.width             = 720,
+	.height            = 480,
+	.field_height      = 240,
+	.aspect_ratio_num  = 4,
+	.aspect_ratio_den  = 3,
+	.sync_duration_num = 60,
+	.sync_duration_den = 1,
+	.video_clk         = 27000000,
+	},
+	{ /* VMODE_480P */
+	.name              = "480p",
+	.mode              = VMODE_480P,
+	.width             = 720,
+	.height            = 480,
+	.field_height      = 480,
+	.aspect_ratio_num  = 4,
+	.aspect_ratio_den  = 3,
+	.sync_duration_num = 60,
+	.sync_duration_den = 1,
+	.video_clk         = 27000000,
+	},
+	{ /* VMODE_480P_RPT */
+	.name              = "480p_rpt",
+	.mode              = VMODE_480P_RPT,
+	.width             = 720,
+	.height            = 480,
+	.field_height      = 480,
+	.aspect_ratio_num  = 4,
+	.aspect_ratio_den  = 3,
+	.sync_duration_num = 60,
+	.sync_duration_den = 1,
+	.video_clk         = 27000000,
+	},
+	{ /* VMODE_576I */
+	.name              = "576i",
+	.mode              = VMODE_576I,
+	.width             = 720,
+	.height            = 576,
+	.field_height      = 288,
+	.aspect_ratio_num  = 4,
+	.aspect_ratio_den  = 3,
+	.sync_duration_num = 50,
+	.sync_duration_den = 1,
+	.video_clk         = 27000000,
+	},
+	{ /* VMODE_576I_RPT */
+	.name              = "576i_rpt",
+	.mode              = VMODE_576I_RPT,
+	.width             = 720,
+	.height            = 576,
+	.field_height      = 288,
+	.aspect_ratio_num  = 4,
+	.aspect_ratio_den  = 3,
+	.sync_duration_num = 50,
+	.sync_duration_den = 1,
+	.video_clk         = 27000000,
+	},
+	{ /* VMODE_576I */
+	.name              = "576cvbs",
+	.mode              = VMODE_576CVBS,
+	.width             = 720,
+	.height            = 576,
+	.field_height      = 288,
+	.aspect_ratio_num  = 4,
+	.aspect_ratio_den  = 3,
+	.sync_duration_num = 50,
+	.sync_duration_den = 1,
+	.video_clk         = 27000000,
+	},
+	{ /* VMODE_576P */
+	.name              = "576p",
+	.mode              = VMODE_576P,
+	.width             = 720,
+	.height            = 576,
+	.field_height      = 576,
+	.aspect_ratio_num  = 4,
+	.aspect_ratio_den  = 3,
+	.sync_duration_num = 50,
+	.sync_duration_den = 1,
+	.video_clk         = 27000000,
+	},
+	{ /* VMODE_576P_RPT */
+	.name              = "576p_rpt",
+	.mode              = VMODE_576P_RPT,
+	.width             = 720,
+	.height            = 576,
+	.field_height      = 576,
+	.aspect_ratio_num  = 4,
+	.aspect_ratio_den  = 3,
+	.sync_duration_num = 50,
+	.sync_duration_den = 1,
+	.video_clk         = 27000000,
+	},
+	{ /* VMODE_720P */
+	.name              = "720p",
+	.mode              = VMODE_720P,
+	.width             = 1280,
+	.height            = 720,
+	.field_height      = 720,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 60,
+	.sync_duration_den = 1,
+	.video_clk         = 74250000,
+	},
 #ifdef CONFIG_AML_VOUT_FRAMERATE_AUTOMATION
-    { /* VMODE_720P_59HZ */
-        .name              = "720p59hz",
-        .mode              = VMODE_720P_59HZ,
-        .width             = 1280,
-        .height            = 720,
-        .field_height      = 720,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 60000,
-        .sync_duration_den = 1001,
-        .video_clk         = 74250000,
-    },
+	{ /* VMODE_720P_59HZ */
+	.name              = "720p59hz",
+	.mode              = VMODE_720P_59HZ,
+	.width             = 1280,
+	.height            = 720,
+	.field_height      = 720,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 60000,
+	.sync_duration_den = 1001,
+	.video_clk         = 74250000,
+	},
 #endif
-    { /* VMODE_1080I */
-        .name              = "1080i",
-        .mode              = VMODE_1080I,
-        .width             = 1920,
-        .height            = 1080,
-        .field_height      = 540,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 60,
-        .sync_duration_den = 1,
-        .video_clk         = 74250000,
-    },
+	{ /* VMODE_1080I */
+	.name              = "1080i",
+	.mode              = VMODE_1080I,
+	.width             = 1920,
+	.height            = 1080,
+	.field_height      = 540,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 60,
+	.sync_duration_den = 1,
+	.video_clk         = 74250000,
+	},
 #ifdef CONFIG_AML_VOUT_FRAMERATE_AUTOMATION
-    { /* VMODE_1080I_59HZ */
-        .name              = "1080i59hz",
-        .mode              = VMODE_1080I_59HZ,
-        .width             = 1920,
-        .height            = 1080,
-        .field_height      = 540,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 60000,
-        .sync_duration_den = 1001,
-        .video_clk         = 74250000,
-    },
+	{ /* VMODE_1080I_59HZ */
+	.name              = "1080i59hz",
+	.mode              = VMODE_1080I_59HZ,
+	.width             = 1920,
+	.height            = 1080,
+	.field_height      = 540,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 60000,
+	.sync_duration_den = 1001,
+	.video_clk         = 74250000,
+	},
 #endif
-    { /* VMODE_1080P */
-        .name              = "1080p",
-        .mode              = VMODE_1080P,
-        .width             = 1920,
-        .height            = 1080,
-        .field_height      = 1080,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 60,
-        .sync_duration_den = 1,
-        .video_clk         = 148500000,
-    },
+	{ /* VMODE_1080P */
+	.name              = "1080p",
+	.mode              = VMODE_1080P,
+	.width             = 1920,
+	.height            = 1080,
+	.field_height      = 1080,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 60,
+	.sync_duration_den = 1,
+	.video_clk         = 148500000,
+	},
 #ifdef CONFIG_AML_VOUT_FRAMERATE_AUTOMATION
-    { /* VMODE_1080P_59HZ */
-        .name              = "1080p59hz",
-        .mode              = VMODE_1080P_59HZ,
-        .width             = 1920,
-        .height            = 1080,
-        .field_height      = 1080,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 60000,
-        .sync_duration_den = 1001,
-        .video_clk         = 148500000,
-    },
+	{ /* VMODE_1080P_59HZ */
+	.name              = "1080p59hz",
+	.mode              = VMODE_1080P_59HZ,
+	.width             = 1920,
+	.height            = 1080,
+	.field_height      = 1080,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 60000,
+	.sync_duration_den = 1001,
+	.video_clk         = 148500000,
+	},
 #endif
-    { /* VMODE_720P_50hz */
-        .name              = "720p50hz",
-        .mode              = VMODE_720P_50HZ,
-        .width             = 1280,
-        .height            = 720,
-        .field_height      = 720,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 50,
-        .sync_duration_den = 1,
-        .video_clk         = 74250000,
-    },
-    { /* VMODE_1080I_50HZ */
-        .name              = "1080i50hz",
-        .mode              = VMODE_1080I_50HZ,
-        .width             = 1920,
-        .height            = 1080,
-        .field_height      = 540,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 50,
-        .sync_duration_den = 1,
-        .video_clk         = 74250000,
-    },
-    { /* VMODE_1080P_50HZ */
-        .name              = "1080p50hz",
-        .mode              = VMODE_1080P_50HZ,
-        .width             = 1920,
-        .height            = 1080,
-        .field_height      = 1080,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 50,
-        .sync_duration_den = 1,
-        .video_clk         = 148500000,
-    },
-    { /* VMODE_1080P_24HZ */
-        .name              = "1080p24hz",
-        .mode              = VMODE_1080P_24HZ,
-        .width             = 1920,
-        .height            = 1080,
-        .field_height      = 1080,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 24,
-        .sync_duration_den = 1,
-        .video_clk         = 74250000,
-    },
+	{ /* VMODE_720P_50hz */
+	.name              = "720p50hz",
+	.mode              = VMODE_720P_50HZ,
+	.width             = 1280,
+	.height            = 720,
+	.field_height      = 720,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 50,
+	.sync_duration_den = 1,
+	.video_clk         = 74250000,
+	},
+	{ /* VMODE_1080I_50HZ */
+	.name              = "1080i50hz",
+	.mode              = VMODE_1080I_50HZ,
+	.width             = 1920,
+	.height            = 1080,
+	.field_height      = 540,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 50,
+	.sync_duration_den = 1,
+	.video_clk         = 74250000,
+	},
+	{ /* VMODE_1080P_50HZ */
+	.name              = "1080p50hz",
+	.mode              = VMODE_1080P_50HZ,
+	.width             = 1920,
+	.height            = 1080,
+	.field_height      = 1080,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 50,
+	.sync_duration_den = 1,
+	.video_clk         = 148500000,
+	},
+	{ /* VMODE_1080P_24HZ */
+	.name              = "1080p24hz",
+	.mode              = VMODE_1080P_24HZ,
+	.width             = 1920,
+	.height            = 1080,
+	.field_height      = 1080,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 24,
+	.sync_duration_den = 1,
+	.video_clk         = 74250000,
+	},
 #ifdef CONFIG_AML_VOUT_FRAMERATE_AUTOMATION
-    { /* VMODE_1080P_23HZ */
-        .name              = "1080p23hz",
-        .mode              = VMODE_1080P_23HZ,
-        .width             = 1920,
-        .height            = 1080,
-        .field_height      = 1080,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 2397,
-        .sync_duration_den = 100,
-        .video_clk         = 74250000,
-    },
+	{ /* VMODE_1080P_23HZ */
+	.name              = "1080p23hz",
+	.mode              = VMODE_1080P_23HZ,
+	.width             = 1920,
+	.height            = 1080,
+	.field_height      = 1080,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 2397,
+	.sync_duration_den = 100,
+	.video_clk         = 74250000,
+	},
 #endif
-    { /* VMODE_4K2K_30HZ */
-        .name              = "4k2k30hz",
-        .mode              = TVMODE_4K2K_30HZ,
-        .width             = 3840,
-        .height            = 2160,
-        .field_height      = 2160,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 30,
-        .sync_duration_den = 1,
-        .video_clk         = 297000000,
-    },
+	{ /* VMODE_4K2K_30HZ */
+	.name              = "4k2k30hz",
+	.mode              = TVMODE_4K2K_30HZ,
+	.width             = 3840,
+	.height            = 2160,
+	.field_height      = 2160,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 30,
+	.sync_duration_den = 1,
+	.video_clk         = 297000000,
+	},
 #ifdef CONFIG_AML_VOUT_FRAMERATE_AUTOMATION
-    { /* VMODE_4K2K_29HZ */
-        .name              = "4k2k29hz",
-        .mode              = TVMODE_4K2K_29HZ,
-        .width             = 3840,
-        .height            = 2160,
-        .field_height      = 2160,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 2997,
-        .sync_duration_den = 100,
-        .video_clk         = 297000000,
-    },
+	{ /* VMODE_4K2K_29HZ */
+	.name              = "4k2k29hz",
+	.mode              = TVMODE_4K2K_29HZ,
+	.width             = 3840,
+	.height            = 2160,
+	.field_height      = 2160,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 2997,
+	.sync_duration_den = 100,
+	.video_clk         = 297000000,
+	},
 #endif
-    { /* VMODE_4K2K_25HZ */
-        .name              = "4k2k25hz",
-        .mode              = TVMODE_4K2K_25HZ,
-        .width             = 3840,
-        .height            = 2160,
-        .field_height      = 2160,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 25,
-        .sync_duration_den = 1,
-        .video_clk         = 297000000,
-    },
-    { /* VMODE_4K2K_24HZ */
-        .name              = "4k2k24hz",
-        .mode              = TVMODE_4K2K_24HZ,
-        .width             = 3840,
-        .height            = 2160,
-        .field_height      = 2160,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 24,
-        .sync_duration_den = 1,
-        .video_clk         = 297000000,
-    },
+	{ /* VMODE_4K2K_25HZ */
+	.name              = "4k2k25hz",
+	.mode              = TVMODE_4K2K_25HZ,
+	.width             = 3840,
+	.height            = 2160,
+	.field_height      = 2160,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 25,
+	.sync_duration_den = 1,
+	.video_clk         = 297000000,
+	},
+	{ /* VMODE_4K2K_24HZ */
+	.name              = "4k2k24hz",
+	.mode              = TVMODE_4K2K_24HZ,
+	.width             = 3840,
+	.height            = 2160,
+	.field_height      = 2160,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 24,
+	.sync_duration_den = 1,
+	.video_clk         = 297000000,
+	},
 #ifdef CONFIG_AML_VOUT_FRAMERATE_AUTOMATION
-    { /* VMODE_4K2K_23HZ */
-        .name              = "4k2k23hz",
-        .mode              = TVMODE_4K2K_23HZ,
-        .width             = 3840,
-        .height            = 2160,
-        .field_height      = 2160,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 2397,
-        .sync_duration_den = 100,
-        .video_clk         = 297000000,
-    },
+	{ /* VMODE_4K2K_23HZ */
+	.name              = "4k2k23hz",
+	.mode              = TVMODE_4K2K_23HZ,
+	.width             = 3840,
+	.height            = 2160,
+	.field_height      = 2160,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 2397,
+	.sync_duration_den = 100,
+	.video_clk         = 297000000,
+	},
 #endif
-    { /* VMODE_4K2K_SMPTE */
-        .name              = "4k2ksmpte",
-        .mode              = TVMODE_4K2K_SMPTE,
-        .width             = 4096,
-        .height            = 2160,
-        .field_height      = 2160,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 24,
-        .sync_duration_den = 1,
-        .video_clk         = 297000000,
-    },
-    { /* VMODE_vga */
-        .name              = "vga",
-        .mode              = VMODE_VGA,
-        .width             = 640,
-        .height            = 480,
-        .field_height      = 240,
-        .aspect_ratio_num  = 4,
-        .aspect_ratio_den  = 3,
-        .sync_duration_num = 60,
-        .sync_duration_den = 1,
-        .video_clk         = 25175000,
-    },
-    { /* VMODE_SVGA */
-        .name              = "svga",
-        .mode              = VMODE_SVGA,
-        .width             = 800,
-        .height            = 600,
-        .field_height      = 600,
-        .aspect_ratio_num  = 4,
-        .aspect_ratio_den  = 3,
-        .sync_duration_num = 60,
-        .sync_duration_den = 1,
-        .video_clk         = 40000000,
-    },
-    { /* VMODE_XGA */
-        .name              = "xga",
-        .mode              = VMODE_XGA,
-        .width             = 1024,
-        .height            = 768,
-        .field_height      = 768,
-        .aspect_ratio_num  = 4,
-        .aspect_ratio_den  = 3,
-        .sync_duration_num = 60,
-        .sync_duration_den = 1,
-        .video_clk         = 65000000,
-    },
-    { /* VMODE_sxga */
-        .name              = "sxga",
-        .mode              = VMODE_SXGA,
-        .width             = 1280,
-        .height            = 1024,
-        .field_height      = 1024,
-        .aspect_ratio_num  = 5,
-        .aspect_ratio_den  = 4,
-        .sync_duration_num = 60,
-        .sync_duration_den = 1,
-        .video_clk         = 108000000,
-    },
+	{ /* VMODE_4K2K_SMPTE */
+	.name              = "4k2ksmpte",
+	.mode              = TVMODE_4K2K_SMPTE,
+	.width             = 4096,
+	.height            = 2160,
+	.field_height      = 2160,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 24,
+	.sync_duration_den = 1,
+	.video_clk         = 297000000,
+	},
+	{ /* VMODE_vga */
+	.name              = "vga",
+	.mode              = VMODE_VGA,
+	.width             = 640,
+	.height            = 480,
+	.field_height      = 240,
+	.aspect_ratio_num  = 4,
+	.aspect_ratio_den  = 3,
+	.sync_duration_num = 60,
+	.sync_duration_den = 1,
+	.video_clk         = 25175000,
+	},
+	{ /* VMODE_SVGA */
+	.name              = "svga",
+	.mode              = VMODE_SVGA,
+	.width             = 800,
+	.height            = 600,
+	.field_height      = 600,
+	.aspect_ratio_num  = 4,
+	.aspect_ratio_den  = 3,
+	.sync_duration_num = 60,
+	.sync_duration_den = 1,
+	.video_clk         = 40000000,
+	},
+	{ /* VMODE_XGA */
+	.name              = "xga",
+	.mode              = VMODE_XGA,
+	.width             = 1024,
+	.height            = 768,
+	.field_height      = 768,
+	.aspect_ratio_num  = 4,
+	.aspect_ratio_den  = 3,
+	.sync_duration_num = 60,
+	.sync_duration_den = 1,
+	.video_clk         = 65000000,
+	},
+	{ /* VMODE_sxga */
+	.name              = "sxga",
+	.mode              = VMODE_SXGA,
+	.width             = 1280,
+	.height            = 1024,
+	.field_height      = 1024,
+	.aspect_ratio_num  = 5,
+	.aspect_ratio_den  = 4,
+	.sync_duration_num = 60,
+	.sync_duration_den = 1,
+	.video_clk         = 108000000,
+	},
 	{ /* VMODE_wsxga */
-        .name              = "wsxga",
-        .mode              = VMODE_WSXGA,
-        .width             = 1440,
-        .height            = 900,
-        .field_height      = 900,
-        .aspect_ratio_num  = 8,
-        .aspect_ratio_den  = 5,
-        .sync_duration_num = 60,
-        .sync_duration_den = 1,
-        .video_clk         = 88750000,
-    },
+	.name              = "wsxga",
+	.mode              = VMODE_WSXGA,
+	.width             = 1440,
+	.height            = 900,
+	.field_height      = 900,
+	.aspect_ratio_num  = 8,
+	.aspect_ratio_den  = 5,
+	.sync_duration_num = 60,
+	.sync_duration_den = 1,
+	.video_clk         = 88750000,
+	},
 	{ /* VMODE_fhdvga */
-        .name              = "fhdvga",
-        .mode              = VMODE_FHDVGA,
-        .width             = 1920,
-        .height            = 1080,
-        .field_height      = 1080,
-        .aspect_ratio_num  = 16,
-        .aspect_ratio_den  = 9,
-        .sync_duration_num = 60,
-        .sync_duration_den = 1,
-        .video_clk         = 148500000,
-    },
+	.name              = "fhdvga",
+	.mode              = VMODE_FHDVGA,
+	.width             = 1920,
+	.height            = 1080,
+	.field_height      = 1080,
+	.aspect_ratio_num  = 16,
+	.aspect_ratio_den  = 9,
+	.sync_duration_num = 60,
+	.sync_duration_den = 1,
+	.video_clk         = 148500000,
+	},
 };
 
 static const struct file_operations am_tv_fops = {
-	.open	= NULL,  
-	.read	= NULL,//am_tv_read, 
-	.write	= NULL, 
-	.unlocked_ioctl	= NULL,//am_tv_ioctl, 
-	.release	= NULL, 	
+	.open	= NULL,
+	.read	= NULL,//am_tv_read,
+	.write	= NULL,
+	.unlocked_ioctl	= NULL,//am_tv_ioctl,
+	.release	= NULL,
 	.poll		= NULL,
 };
 
@@ -543,7 +543,7 @@ static const vinfo_t *get_valid_vinfo(char  *mode)
 	const vinfo_t * vinfo = NULL;
 	int  i,count=ARRAY_SIZE(tv_info);
 	int mode_name_len=0;
-	
+
 	for(i=0;i<count;i++)
 	{
 		if(strncmp(tv_info[i].name,mode,strlen(tv_info[i].name))==0)
@@ -581,10 +581,10 @@ static int tv_set_current_vmode(vmode_t mod)
 static vmode_t tv_validate_vmode(char *mode)
 {
 	const vinfo_t *info = get_valid_vinfo(mode);
-	
+
 	if (info)
 		return info->mode;
-	
+
 	return VMODE_MAX;
 }
 static int tv_vmode_is_supported(vmode_t mode)
@@ -624,19 +624,19 @@ static int tv_resume(int pm_event)
 	tv_set_current_vmode(info->vinfo->mode);
 	return 0;
 }
-#endif 
+#endif
 static vout_server_t tv_server={
 	.name = "vout2_tv_server",
-	.op = {	
+	.op = {
 		.get_vinfo=tv_get_current_info,
 		.set_vmode=tv_set_current_vmode,
 		.validate_vmode=tv_validate_vmode,
 		.vmode_is_supported=tv_vmode_is_supported,
 		.disable = tv_module_disable,
-#ifdef  CONFIG_PM  
+#ifdef  CONFIG_PM
 		.vout_suspend=tv_suspend,
 		.vout_resume=tv_resume,
-#endif	
+#endif
 	},
 };
 
@@ -650,15 +650,15 @@ static vout_server_t tv_server={
 **	The 6th digit control s-video chroma output DAC number
 ** 	examble :
 **		echo  120120 > /sys/class/display/vdac_setting
-**		the first digit from the left side .	
+**		the first digit from the left side .
 ******************************************************/
-static void  parse_vdac_setting(char *para) 
+static void  parse_vdac_setting(char *para)
 {
 	int  i;
 	char  *pt=strstrip(para);
 	int len=strlen(pt);
 	u32  vdac_sequence=get_current_vdac_setting2();
-	
+
 	amlog_mask_level(LOG_MASK_PARA,LOG_LEVEL_LOW,"origin vdac setting:0x%x,strlen:%d\n",vdac_sequence,len);
 	if(len!=6)
 	{
@@ -673,7 +673,7 @@ static void  parse_vdac_setting(char *para)
 		pt++;
 	}
 	amlog_mask_level(LOG_MASK_PARA,LOG_LEVEL_LOW,"current vdac setting:0x%x\n",vdac_sequence);
-	
+
 	change_vdac_setting2(vdac_sequence,get_current_vmode2());
 }
 static  struct  class_attribute   *tv_attr[]={
@@ -701,29 +701,30 @@ static int  create_tv_attr(disp_module_info_t* info)
 	sprintf(vdac_setting,"%x",get_current_vdac_setting2());
 	return   0;
 }
-static int __init tv_init_module(void)
+
+static int tvout_probe(struct platform_device *pdev)
 {
 	int  ret ;
 
-	info=(disp_module_info_t*)kmalloc(sizeof(disp_module_info_t),GFP_KERNEL) ;
+	info = (disp_module_info_t*)kmalloc(sizeof(disp_module_info_t),GFP_KERNEL) ;
 	if (!info)
 	{
 		amlog_mask_level(LOG_MASK_INIT,LOG_LEVEL_HIGH,"can't alloc display info struct\n");
 		return -ENOMEM;
 	}
-	
+
 	memset(info, 0, sizeof(disp_module_info_t));
 
 	sprintf(info->name,TV_CLASS_NAME) ;
 	ret=register_chrdev(TV2_CONF_MAJOR,info->name,&am_tv_fops);
-	if(ret <0) 
+	if (ret < 0)
 	{
 		amlog_mask_level(LOG_MASK_INIT,LOG_LEVEL_HIGH,"register char dev tv error\n");
 		return  ret ;
 	}
 	info->major=TV2_CONF_MAJOR;
 	amlog_mask_level(LOG_MASK_INIT,LOG_LEVEL_HIGH,"major number %d for disp\n",ret);
-	if(vout2_register_server(&tv_server))
+	if (vout2_register_server(&tv_server))
 	{
 		amlog_mask_level(LOG_MASK_INIT,LOG_LEVEL_HIGH,"register tv module server fail\n");
 	}
@@ -735,32 +736,72 @@ static int __init tv_init_module(void)
 	return 0;
 
 }
-static __exit void tv_exit_module(void)
+
+static int __exit tvout_remove(struct platform_device *pdev)
 {
 	int i;
-	
-	if(info->base_class)
+
+	if (info->base_class)
 	{
-		for(i=0;i<ARRAY_SIZE(tv_attr);i++)
+		for (i=0; i<ARRAY_SIZE(tv_attr); i++)
 		{
 			class_remove_file(info->base_class,tv_attr[i]) ;
 		}
-		
+
 		class_destroy(info->base_class);
-	}	
-	if(info)
+	}
+	if (info)
 	{
 		unregister_chrdev(info->major,info->name)	;
 		kfree(info);
 	}
 	vout2_unregister_server(&tv_server);
-	
+
 	amlog_mask_level(LOG_MASK_INIT,LOG_LEVEL_HIGH,"exit tv module\n");
+	return 0;
 }
 
 
-arch_initcall(tv_init_module);
-module_exit(tv_exit_module);
+#ifdef CONFIG_USE_OF
+static const struct of_device_id tvout_dt_match[] = {
+	{
+		.compatible = "amlogic,tvout2",
+	},
+	{},
+};
+#endif
+
+static struct platform_driver tvout_driver = {
+	.driver = {
+		.name = TV_CLASS_NAME,
+		.owner = THIS_MODULE,
+#ifdef CONFIG_USE_OF
+		.of_match_table = tvout_dt_match,
+#endif
+	},
+	.probe = tvout_probe,
+	.remove = __exit_p(tvout_remove),
+};
+
+static int __init tvout_init_module(void)
+{
+	printk("%s module init\n", __func__);
+	if (platform_driver_register(&tvout_driver)) {
+		printk("%s failed to register module\n", __func__);
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static void __exit tvout_exit_module(void)
+{
+	printk("%s module exit\n", __func__);
+	platform_driver_unregister(&tvout_driver);
+}
+
+arch_initcall(tvout_init_module);
+module_exit(tvout_exit_module);
 
 MODULE_DESCRIPTION("display configure  module");
 MODULE_LICENSE("GPL");
-- 
2.19.0

